Add new zephyr workflows and documentation (#7)
IRNAS/irnas-workflows-software: workflow-templates/zephyr/README.md

# Zephyr

## Description

Workflows in this group extend the workflows in [Basic](../basic/README.md)
group and introduce several new functionalities which aid in the development of
the projects using Zephyr.

Specifically, they provide:

- Everything that [Basic](../basic/README.md) workflows already do: changelog
  preparation, tagging, publishing GitHub releases, etc.,
- Running _builds_ on Pull Requests and during releases processes,
- Caching of West modules and toolchains downloaded by East to speed up the
  _build_ process,
- A way to change what _build_ means for each project,

> - A way to specify which artefacts are attached to the published GitHub releases

Reviewer: You are using artefacts and artifacts throughout this document, stick to one for consistency

Needed files (relative to project's root dir):

- `scripts/requirements.txt`, can be empty,
- `makefile` - Specific content is expected, see section
  [How to configure _build_](#how-to-configure-build).

### How to use

This group contains four workflows:

- `create-release.yaml`
- `publish-release.yaml`
- `build.yaml`
- `label_pr.yaml`

They can be used in two different scenarios:

- During a release process
- In Pull Requests

#### Release process

To trigger a release process just follow the instructions in the Basic's
[How to use](../basic/README.md#how-to-use) section.

Everything that is described in that section still applies, with some
modifications:

- After creating a release tag and a new changelog section a `build.yaml`
  workflow is called.
- `build.yaml` runs a _build_ process.
- After that `publish-release.yaml` takes any resulting _build_ artifacts and
  creates a GitHub release with them.

If anything goes wrong during `build.yaml` and `publish-release.yaml` workflows
then the created release tag and Changelog update commit are deleted from the
`main` branch.

#### Pull requests

`build.yaml` (and therefore _build_ process) is automatically triggered whenever

> a PR is opened, reopened or a new commit is pushed to it.

Reviewer: It triggers when a commit is pushed to the existing PR, right? Try wording this a bit more clearly.

Make and the `makefile` file present in the project's root directory.

After running the equivalent of the below commands:

```bash
# Clone the repo into the `project` folder
# mkdir -p <project_name>/project
# cd <project_name>/project
# git clone <project_url> .

# Check the cache for West modules and East toolchain and extract if found
# Check the cache for Python dependencies based on scripts/requirements.txt
```

the `build.yaml` starts to execute `make` commands in the following order:

```bash
make install-dep
make project-setup
make pre-build
make build
make pre-package    # Used in the release process, skipped during PRs
```

It is up to the developer to decide what these commands do. A good starting
point is the
[makefile](https://github.com/IRNAS/irnas-zephyr-template/blob/main/makefile)
that is provided by `irnas-zephyr-template` repo by default.

Note: all `make` commands are executed from the root of the cloned repository.
Even if some command would `cd` into some other folder, the next command would
still execute from the root.

Expected behaviour of each `make` command:

- `make install-dep` - Installs tooling needed by the project.

> - `make project-setup` - Set-ups the project and any tooling that might depend

Reviewer: Sets up

  on it.
- `make pre-build` - Runs commands that need to run before the _build_.
- `make build` - The _build_, the heart of the workflow. This could be a single
  build command or a set of build commands.
- `make pre-package` - Only used in the release process, it is skipped if the
  workflow was triggered due to a PR. Use it to package build artefacts.

### Packaging build artefacts

In release processes, the `build.yaml` will collect any files found in the
`artefacts` folder of the project's root directory and attach them to the newly
created a GitHub release as release artefacts.

> The developer can use `make pre-package` command to create the `artefacts`

Reviewer: can use the `make pre-package`...

folder and move any files of interest inside it.

Packaging of build artefacts is done only in release processes, it is skipped if
the workflow was triggered due to a PR.

### Adding extra text to the Release notes

Some projects need additional information in the Release Notes apart from the
changelog notes, maybe general usage instructions, an explanation of the build
artefacts or some dynamically generated report.

To do that the `make pre-package` command can copy into the `artefacts` folder
the `pre_changelog.md` and `post_changelog.md` markdown files.

Contents of these files then becomes a part of the Release Notes in the
following way:

```markdown
# Release notes

{Contents of the pre_changelog.md}

{Contents of the changelog for this version}

{Contents of the post_changelog.md}
```

> If some section is not used, its file can be empty. `pre_changelog.md` and

Reviewer 2: ```suggestion
If a section is not used, the corresponding file can be empty. `pre_changelog.md` and
```
